spanner/regression_test.go

```diff
@@ -0,0 +1,214 @@
+// limitations under the License.
+
+package spanner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"maps"
+	"slices"
+	"sort"
+	"testing"
+
+	sppb "cloud.google.com/go/spanner/apiv1/spannerpb"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/iterator"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"cloud.google.com/go/spanner/internal/testutil"
+)
+
+type methodAndMetadata struct {
+	method string
+	md     metadata.MD
+}
+
+type ourInterceptor struct {
+	unaryHeaders  []*methodAndMetadata
+	streamHeaders []*methodAndMetadata
+}
+
+func (oi *ourInterceptor) interceptStream(srv any, ss grpc.ServerStream, ssi *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	md, ok := metadata.FromIncomingContext(ss.Context())
+	if !ok {
+		return errors.New("missing metadata in stream")
+	}
+	oi.streamHeaders = append(oi.streamHeaders, &methodAndMetadata{ssi.FullMethod, md})
+	return handler(srv, ss)
+}
+
+func (oi *ourInterceptor) interceptUnary(ctx context.Context, req any, usi *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, errors.New("missing metadata in unary")
+	}
+	oi.unaryHeaders = append(oi.unaryHeaders, &methodAndMetadata{usi.FullMethod, md})
+	return handler(ctx, req)
+}
+
+// This is a regression test to assert that all the expected headers are propagated
+// along to the final gRPC server avoiding scenarios where headers got dropped from a
+// destructive context augmentation call.
+// Please see https://github.com/googleapis/google-cloud-go/issues/11656
+func TestAllHeadersForwardedAppropriately(t *testing.T) {
+	if isMultiplexEnabled {
+		t.Skip("Skipping these tests with multiplexed sessions until #11308 is fixed")
+	}
+
+	// 1. Set up the server interceptor that'll record and collect
+	// all the headers that  are received by the server.
+	oint := new(ourInterceptor)
+	sopts := []grpc.ServerOption{
+		grpc.UnaryInterceptor(oint.interceptUnary), grpc.StreamInterceptor(oint.interceptStream),
+	}
+	mockedServer, clientOpts, teardown := makeMockServer(t, sopts...)
+	defer teardown()
+
+	clientConfig := ClientConfig{
+		SessionPoolConfig: SessionPoolConfig{
+			MinOpened:     2,
+			MaxOpened:     10,
+			WriteSessions: 0.2,
+			incStep:       2,
+		},
+		EnableEndToEndTracing: true,
+		DisableRouteToLeader:  false,
+	}
+	formattedDatabase := fmt.Sprintf("projects/%s/instances/%s/databases/%s", "[PROJECT]", "[INSTANCE]", "[DATABASE]")
+	sc, err := NewClientWithConfig(context.Background(), formattedDatabase, clientConfig, clientOpts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer sc.Close()
+
+	// 2. Perform a simple "SELECT 1" to trigger both unary and streaming gRPC calls.
+	sqlSELECT1 := "SELECT 1"
+	resultSet := &sppb.ResultSet{
+		Rows: []*structpb.ListValue{
+			{Values: []*structpb.Value{
+				{Kind: &structpb.Value_NumberValue{NumberValue: 1}},
```

Comment:
nit: not directly relevant to this test, as it does not read the value, but this encoding is not correct. INT64 values are encoded as string values. NumberValue is only used for FLOAT32 and FLOAT64. But it would be good to fix that here, as others might copy-paste this test later and run into unexpected errors when they try to read the data.
(If you want to verify that it fails: Try to do a `row.Scan(..)` in the loop below that iterates over the rows.)

Reply:
Gotcha and thanks, done! Please take a look.